src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,256 @@
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate long Control(ref bio_st bio, BioControl cmd, long num, void* ptr);
+        [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
+        private delegate int Free(ref bio_st data);
+
+        private static class WindowsLib
+        {
+            public const string CryptoDll = "libeay32.dll";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static class UnixLib
+        {
+            public const string CryptoDll = "libcrypto.so";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static class OsxLib
+        {
+            public const string CryptoDll = "libcrypto.1.0.0.dylib";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static void BIO_set_flags(ref bio_st bio, BioFlags flags)
+        {
+            if (Interop.IsWindows)
+            {
+                WindowsLib.BIO_set_flags(ref bio, flags);
+            }
+            else if (Interop.IsOsx)
+            {
+                OsxLib.BIO_set_flags(ref bio, flags);
+            }
+            else
+            {
+                UnixLib.BIO_set_flags(ref bio, flags);
+            }
+        }
+
+        private static int FreeBio(ref bio_st bio)
+        {
+            if (bio.init == 1 && bio.shutdown == 1)
+            {
+                bio.ptr = null;
+                return 1;
+            }
+            return -1;
+        }
+
+        private static int CreateBio(ref bio_st bio)
+        {
+            bio.shutdown = 1;
+            bio.init = 1;
+            bio.num = -1;
+            return 1;
+        }
+
+        private static int WriteBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var bufferHandle = GCHandle.FromIntPtr((IntPtr)bio.ptr);
+            var bufferPipeline = (IPipelineWriter)bufferHandle.Target;
+            int numberOfBytesRemaing = numberOfBytes;
+            var buffer = bufferPipeline.Alloc();
+            while (numberOfBytesRemaing > 0)
+            {
+                int sizeToWrite = Math.Min(MaxBlockSize, numberOfBytesRemaing);
+                buffer.Ensure(sizeToWrite);
+                var writeSpan = new Span<byte>(buff, sizeToWrite);
+                writeSpan.CopyTo(buffer.Memory.Span);
+                buffer.Advance(sizeToWrite);
+                numberOfBytesRemaing -= sizeToWrite;
+            }
+            buffer.Commit();
+            bio.num_write = (uint)numberOfBytes;
+            BIO_set_flags(ref bio, BioFlags.BIO_FLAGS_WRITE | BioFlags.BIO_FLAGS_SHOULD_RETRY);
```

Comment:
I need to recheck, but I stretching my memory back, there was an issue without it..... OpenSSL doc are.... not always right.